### PR TITLE
[Enhancement] Radio & Checkbox Icon Alignment

### LINKED ISSUE #1359: [Enhancement] Radio-icon & Checkbox-icon alignment
**Please add a short description of your enhancement request**
When a Kirby-radio label spans over more than one line - the radio-icon should be top aligned.

**Describe the solution you'd like**
Align the radio-icon with the first line of the label.

**Describe alternatives you've considered**
Let the radio-icon be center aligned AS-IS - but this results in layout that does not match the designed layout.

**Additional context**

Radio-Issue

### Kick Off:

- [X] Ensure the enhancement is `UX refined` and aligned with UX 
_The component and/or enhancement should be published and available in the [Kirby Styleguide on Zeplin](https://zpl.io/258pXGj)_
- [X] Ensure the enhancement has been `Tech refined` with @kirbydesign/kirby-guild and this issue is updated with a clear implementation description  
_This issue should be in the [Ready to do](https://github.com/kirbydesign/designsystem/projects/1#column-4590936) column of the [Kirby kan-ban board](https://github.com/kirbydesign/designsystem/projects/1) before starting implementation)_
- [X] Assign yourself to this issue and move it to the [In progress](https://github.com/kirbydesign/designsystem/projects/1#column-4590937) column of the [Kirby kan-ban board](https://github.com/kirbydesign/designsystem/projects/1)

### Code:

- [X] Create Feature Branch from [master branch](https://github.com/kirbydesign/designsystem/tree/master)
- [X] Create a draft implementation and push to Github
- [X] Ask a member of @kirbydesign/kirby-guild for a WIP review by creating a draft Pull Request
- [X] Implement unit tests
- [X] Update Cookbook Examples and Showcase, i.e. see [Radio](https://cookbook.kirby.design/home/showcase/radio)  
_Also remember to add any relevant new API documentation_

### Review:

- UX review:
  - [X] Ensure implementation is correct in relation to the UX design and the [Kirby Styleguide on Zeplin](https://zpl.io/258pXGj)
  - [X] With UX agree on the version of the implementation
- Code review:
  - [X] Open a pull request (or mark the existing draft PR as `Ready for review`) and ask @kirbydesign/kirby-guild for a review  
_Remember to add `closes #issueno` to the description of the PR._
  - [ ] Once approved, merge feature branch/PR to master
- [ ] Ask a member of @kirbydesign/kirby-guild to add a link to component showcase from Kirby Component Status and update the version number

:tada: Celebrate

### PR DESCRIPTION
**Pull request by MadsBuchmann**
This PR closes #1359

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tests for the changes have been added (for bug fixes / features) 
- [X] Docs have been added / updated (for bug fixes / features) ~~(I do not think it is necessary for this change)~~ edit: reviewer begs to differ 😉)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [X] Enhancement (to existing content)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #1359 

## What is the new behavior?
The `checkbox`- and `radio`-component will now have their icon aligned with the first line of the label-text when there's multiple lines in the label.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Other information
I had to change the `beforeEach()` in `checkbox.component.spec.ts` as the actual checkbox was never actually being rendered, causing tests to fail. 

## Screenshots 
Screenshot 2021-03-02 at 13 11 49
Screenshot 2021-03-02 at 13 12 59

### CODE REVIEW

**Comment by MadsBuchmann** [comment]
I have paused work on issue #1359 until i have solved #1378. The latter issue arose from UX-review of this pull-request and i want to make sure text is rendered correctly before continuing work on this issue 😊

**Comment by MadsBuchmann** [comment]
Blocked - see: https://github.com/kirbydesign/designsystem/pull/1374#issuecomment-790513346

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.spec.ts [hunk 1]
```diff
@@ -14,17 +16,18 @@ const checkboxSizeSm = fatFingerSize;
...
 
 describe('CheckboxComponent', () => {
-  let spectator: Spectator<CheckboxComponent>;
-  let ionCheckbox: HTMLIonCheckboxElement;
-
-  const createHost = createHostFactory({
+  const createComponent = createComponentFactory({
     component: CheckboxComponent,
-    declarations: [IonCheckbox],
+    imports: [IonicModule.forRoot({ mode: 'ios', _testing: true })],
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/checkbox/checkbox.component.spec.ts:21]: Is this change relevant to the issue at hand? Do you need to import the entire Ionic module?

File: libs/designsystem/src/lib/components/radio/radio.component.scss [hunk 2]
```diff
@@ -6,26 +6,32 @@ $radio-icon-mark-size: 60%;
 $spacing-to-edge: map-get($checkbox-radio-spacing, 'to-edge');
 $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
 
+@mixin wrapper-top-bottom-padding($target-height) {
+  $padding: calc((#{$target-height} - #{$radio-icon-size}) / 2);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/radio/radio.component.scss:10]: Do you need to calculate this at runtime? If not, consider replacing with a built time calc:
```suggestion
  $padding: ($target-height - $radio-icon-size) / 2;
```

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss [hunk 3]
```diff
@@ -5,6 +5,13 @@ $checkbox-icon-size: size('m');
 $spacing-to-edge: map-get($checkbox-radio-spacing, 'to-edge');
 $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
 
+@mixin wrapper-top-bottom-padding($target-height) {
+  $padding: calc((#{$target-height} - #{$checkbox-icon-size}) / 2);
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:9]: Do you need to calculate this at runtime? If not, consider replacing with a built time calc:
```suggestion
  $padding: ($target-height - $checkbox-icon-size) / 2;
```

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss [hunk 4]
```diff
@@ -80,17 +87,17 @@ $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
     .wrapper {
       position: relative;
       display: inline-flex;
-      align-items: center;
+      align-items: flex-start;
       vertical-align: top;
       padding-right: $spacing-to-label; // Ensure same padding on right side of label
-      min-height: map-get($checkbox-radio-sizes, 'md');
+      @include wrapper-top-bottom-padding(map-get($checkbox-radio-sizes, 'md'));
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:93]: Consider inlining mixin for readability (and move above `padding-right`):
```suggestion
      padding: (map-get($checkbox-radio-sizes, 'md') - $checkbox-icon-size) / 2;
```

File: libs/designsystem/src/lib/components/radio/radio.component.scss [hunk 5]
```diff
@@ -6,26 +6,32 @@ $radio-icon-mark-size: 60%;
...
 
     div.wrapper {
       position: relative;
       display: inline-flex;
-      align-items: center;
+      align-items: flex-start;
       vertical-align: top;
       padding-right: $spacing-to-label; // Ensure same padding on right side of label
-      min-height: map-get($checkbox-radio-sizes, 'md');
+      @include wrapper-top-bottom-padding(map-get($checkbox-radio-sizes, 'md'));
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/radio/radio.component.scss:27]: Consider inlining mixin for readability (and move above `padding-right`):
```suggestion
      padding: (map-get($checkbox-radio-sizes, 'md') - $checkbox-icon-size) / 2;
```

Comment (jakobe) [inline libs/designsystem/src/lib/components/radio/radio.component.scss:27]: Consider extracting `map-get($checkbox-radio-sizes, 'md')` to named variable:
`$default-checkbox-radio-size: map-get($checkbox-radio-sizes, 'md');`

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 4 above)
Comment (jakobe) [inline libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:93]: Consider extracting `map-get($checkbox-radio-sizes, 'md')` to named variable:
`$default-checkbox-radio-size: map-get($checkbox-radio-sizes, 'md');`

**Review by jakobe — CHANGES_REQUESTED** [review]

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 3 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:9]: Good point!

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.spec.ts (hunk 1 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.spec.ts:21]: The `ion-checkbox` was not being rendered during testing in Chrome, which caused tests to fail when relying on padding to maintain the height of the component. 

As the `radio`-component is similar and its slotted `ion-radio` is actually being rendered during testing in chrome; I had a look at what was done in `radio.component.spec.ts` and (perhaps a bit too) trustfully did it the same way - causing `ion-checkbox` to actually be rendered 😄

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 3 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss]: Good point! - Should be fixed by d28acfae53c7e90e4c133f64bb70af305339fa72

File: libs/designsystem/src/lib/components/radio/radio.component.scss (hunk 2 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio.component.scss]: Good point! - Should be fixed by d28acfae53c7e90e4c133f64bb70af305339fa72

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 4 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss]: Done 😀 e9f9de5

File: libs/designsystem/src/lib/components/radio/radio.component.scss (hunk 5 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio.component.scss]: Done 😀 e9f9de5

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 4 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss]: Resolved by ee8237b

File: libs/designsystem/src/lib/components/radio/radio.component.scss (hunk 5 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio.component.scss]: Resolved by: 4e49a6f

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss [hunk 6]
```diff
@@ -107,6 +107,11 @@ $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
       flex-shrink: 0;
       flex-grow: 0;
     }
+
+    span {
+      align-self: center;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:112]: Why is this necessary? 🤔

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss [hunk 7]
```diff
@@ -107,6 +107,11 @@ $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
...
+      line-height: line-height('n');
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:113]: Should the line-height be set on the entire comp instead? 🤔

File: libs/designsystem/src/lib/components/radio/radio.component.scss [hunk 8]
```diff
@@ -41,6 +41,11 @@ $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
       margin-left: $spacing-to-edge;
       margin-right: $spacing-to-label;
     }
+
+    span {
+      align-self: center;
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/radio/radio.component.scss:46]: Why is this necessary? 🤔

File: libs/designsystem/src/lib/components/radio/radio.component.scss [hunk 9]
```diff
@@ -41,6 +41,11 @@ $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
...
+      line-height: line-height('n');
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/radio/radio.component.scss:47]: Should the line-height be set on the entire comp instead? 🤔

**Review by jakobe — COMMENTED** [review]
Consider adding example for checkbox+radio with multiple line labels

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 6 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:112]: When I initially set `align-self: center` line-height was not set and the label was not properly center-aligned with the actual radio/checkbox:
Screenshot 2021-03-11 at 10 09 13

After applying `align-self: center` the label was properly aligned: 
Screenshot 2021-03-11 at 10 11 56

Now that you point it out I can however see that `align-self: center` is no longer necessary due to the line-height matching the radio/checkbox - so I will go ahead and remove it 😄

File: libs/designsystem/src/lib/components/radio/radio.component.scss (hunk 8 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio.component.scss:46]: https://github.com/kirbydesign/designsystem/pull/1374#discussion_r592188753

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss [hunk 10]
```diff
@@ -76,21 +81,25 @@ $spacing-to-label: map-get($checkbox-radio-spacing, 'to-label');
 
   &.has-label {
     display: block;
+    line-height: line-height('n');
+    white-space: pre-line;
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:85]: I've taken the liberty of adding this property to allow for `\n` and `&#10;` which could be usable when working with multiline labels.

File: libs/designsystem/src/lib/components/radio/radio.component.scss [hunk 11]
```diff
@@ -5,27 +5,36 @@ $radio-icon-size: size('m');
...
+  @return ($target-height - $radio-icon-size) / 2;
+}
 
 :host {
   display: inline-block;
 
   &.has-label {
     display: block;
+    line-height: line-height('n');
+    white-space: pre-line;
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/radio/radio.component.scss:20]: see: https://github.com/kirbydesign/designsystem/pull/1374/files#r592378940

File: libs/designsystem/src/lib/components/radio/radio.component.scss (hunk 9 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio.component.scss]: Should be resolved by: a8e1e31

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 6 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss]: Should be resolved by: a8e1e31

File: libs/designsystem/src/lib/components/radio/radio.component.scss (hunk 8 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio.component.scss]: resolved by: a8e1e31

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 7 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss]: resolved by: a8e1e31

**Comment by MadsBuchmann** [comment]
Cookbook example has been added for both checkbox and radio 😄

File: libs/designsystem/src/lib/components/checkbox/checkbox.component.scss (hunk 10 above)
Comment (jakobe) [reply libs/designsystem/src/lib/components/checkbox/checkbox.component.scss:85]: Good thinking @MadsBuchmann 👌🏻

**Review by jakobe — APPROVED** [review]
Good work! 🎉
